[Logging Improvement] Using lambda invocations instead of checking debug/trace isEnabled explicitly

akolarkunnu opened the pull request:
Converted debug/trace/warn/info checks to lambda based logging APIs.

Present scenario:
if (logger.isTraceEnabled()) {
    logger.trace("Some long-running operation returned {}", expensiveOperation());
}
Replace the existing checking of trace/debug logging before actual logging with lambda invocations within the debug/trace loggers:
logger.trace("Some long-running operation returned {}", () -> expensiveOperation());

This achieves the same lazy logging without having to check for logger levels. 

Javadoc of "void debug(String message, Supplier<?>... paramSuppliers)" -> "Logs a message with parameters which are only to be constructed if the logging level is the DEBUG level."

Resolves #8646


### Check List
- [] ~New functionality includes testing.~
- [x] All tests pass
- [ ] ~New functionality has been documented.~
- [] ~New functionality has javadoc added~
- [ ] ~API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).~
- [ ] ~Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))~
- [x] Commits are signed per the DCO using --signoff
- [] ~Commit changes are listed out in CHANGELOG.md file (See: [Changelog]~(../blob/main/CONTRIBUTING.md#changelog))~
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

akolarkunnu committed:
[Logging Improvement] Using lambda invocations instead of checking debug/trace isEnabled explicitly

Converted debug/trace/warn/info checks to lambda based logging APIs.

Resolves #8646

Signed-off-by: Abdul Muneer Kolarkunnu <muneer.kolarkunnu@netapp.com>

vikasvb90 commented on modules/reindex/src/main/java/org/opensearch/index/reindex/AbstractAsyncBulkByScrollAction.java:
```diff
@@ -403,14 +403,12 @@ void prepareBulkRequest(long thisBatchStartTimeNS, ScrollableHitSource.AsyncResp
-                new ByteSizeValue(request.estimatedSizeInBytes())
-            );
-        }
+        logger.debug(
```
@akolarkunnu Thanks for raising this PR! From code readability and simplicity point of view your improvement definitely looks better but not having the if block will now lead to creation of new string and runnable objects. Runnable objects can be lightweight depending upon whether they are referencing any variables outside of their scope but string objects may or may not be lightweight and they may or may not be created every time.
I am not in strong disagreement of this change but I would like to see how JVM and CPU graphs look like with some benchmarks before and after this change.

akolarkunnu replied:
Hi @vikasvb90 This change achieves the same lazy logging without having to check for logger levels.
Javadoc of "void debug(String message, Supplier<?>... paramSuppliers)"  says "Logs a message with parameters which are only to be constructed if the logging level is the DEBUG level."
Are you saying to run any specific existing Microbenchmark test case or write a new Microbenchmark test case which covers this changed code area and execute that?

vikasvb90 replied:
The message is very clear that only the log message is constructed later but the arguments which you are now passing will still occupy space in memory and most of the space would be occupied by string objects. So, you need to first check in total what is the impact of this change on heap. 

I am not referring to micro benchmarks but macro benchmarks. https://github.com/opensearch-project/opensearch-benchmark

sandeshkr419 replied:
Hi @vikasvb90 

I don't think we will see any quantifiable impact in benchmarking because:
1. Overhead of logging is considerably less than indexing/search activities against where we will benchmark them.
2. In our existing code, we already check for isDebugEnabled/isTraceEnabled at most places, so the code path is already short-circuited. It is the only extra boolean check of isDebugEnabled/isTraceEnabled that we will avoid. 

If we had the debug/trace logs without the boolean isDebugEnabled/isTraceEnabled checks, we may have seen some improvement, but in this case, we will not. This refactoring is intended more from a cleaner code and aiming towards best coding practices, and does not contribute significantly towards performance gains.

vikasvb90 replied:
I am not talking about the overhead of logging itself but the attempt of logging which gets discarded later. I understand what it is intended for. To reiterate, boolean checks earlier prevented the construction of objects passed in log but now they will start occupying space and compute as a result of their construction. Also, it isn't about how big the impact is as compared to search or indexing but whether there is any impact at all and I believe there is. 
If you really want this change to be pushed, at least lets first correctly measure the impact of this change on heap.

sandeshkr419 replied:
@akolarkunnu I think some benchmarking results with profiling visualizations will help establish confidence on the changes although this is recommended by log4j: https://logging.apache.org/log4j/2.x/manual/api.html#java-8-lambda-support-for-lazy-logging

We can probably run `geonames` workload on a single node cluster with these configutations:
- no changes, debug & trace disabled
- your changes, debug & trace disabled
AND
- no changes, debug & trace enabled
- your changes, debug & trace enabled


Alternatively, you can reuse the exercise I pointed in https://github.com/opensearch-project/OpenSearch/pull/14723#issuecomment-2289839653 as well - that should also get you CPU profiling. AFAIK, I know ClusterState/Metadata classes have considerate amount of debug/trace logs - verify once.

akolarkunnu committed:
[Logging Improvement] Using lambda invocations instead of checking debug/trace isEnabled explicitly

Converted debug/trace/warn/info checks to lambda based logging APIs.

Resolves #8646

Signed-off-by: Abdul Muneer Kolarkunnu <muneer.kolarkunnu@netapp.com>

akolarkunnu commented:
Why are these precommit tasks are not executing? Any issues from CI?

sandeshkr419 commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -1192,50 +1192,45 @@ private Engine.IndexResult index(Engine engine, Engine.Index index) throws IOExc
         active.set(true);
         final Engine.IndexResult result;
         index = indexingOperationListeners.preIndex(shardId, index);
+        final Engine.Index finalIndex = index;
```
How about:

`final Engine.Index finalIndex = indexingOperationListeners.preIndex(shardId, index);`

and then use `finalIndex` throughout.

also, do take into account that we may be breaking some things here because in L1234:
` indexingOperationListeners.postIndex(shardId, index, e);` because this operation will need to be operated on finalIndex then.

akolarkunnu replied:
Fixed it

sandeshkr419 commented on server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:
```diff
@@ -367,6 +365,8 @@ private void balanceByWeights() {
                 final BalancedShardsAllocator.ModelNode maxNode = modelNodes[highIdx];
                 advance_range: if (maxNode.numShards(index) > 0) {
                     final float delta = absDelta(weights[lowIdx], weights[highIdx]);
+                    final int finalHighIdx = highIdx;
```
Instead of creating new final variables to be used, what about creating 2 private utilities like `log_stop_balancing/balancing(final variables...)` and then just call those utilities instead. I don't want us to create noise changes in code logic.

akolarkunnu replied:
If I understood you correctly, you are suggesting like having a method like below for "lowIdx"

final int getLowIndex(final int index) {
        return index;
    }

and invoking this method from lambda expression like "getLowIndex(lowIdx)"
This won't be possible, because again "lowIdx" is non-final and we can not refer that from lambda expression.
If you meant something else please correct me.

sandeshkr419 replied:
No, abstracting out entire logging statement itself.

```
private log_stop_balancing(final int highIdx, final int lowIdx) {
     logger.trace(
                                 "Stop balancing index [{}]  min_node [{}] weight: [{}]" + "  max_node [{}] weight: [{}]  delta: [{}]",
                            () -> index,
                            () -> maxNode.getNodeId(),
                            () -> weights[finalHighIdx],
                            () -> minNode.getNodeId(),
                            () -> weights[finalLowIdx],
                            () -> delta
}

private log_balancing(final int highIdx, final int lowIdx) {
...
}
```

and then just calling this utility method.

sandeshkr419 commented:
> Why are these precommit tasks are not executing? Any issues from CI?

@akolarkunnu I think its probably because of merge conflicts. We can check CI once again once you resolve conflicts and address the code comments.

akolarkunnu commented on server/src/main/java/org/opensearch/cluster/routing/allocation/decider/DiskThresholdDecider.java:
```diff
@@ -298,15 +294,13 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
-                }
+                logger.debug(
+                    "less than the required {} free bytes threshold ({} free) on node {}, "
+                        + "but allowing allocation because primary has never been allocated",
```
If we remove + here, it affects code readability, it will become ~160 characters per one line.

akolarkunnu committed:
[Logging Improvement] Using lambda invocations instead of checking debug/trace isEnabled explicitly

Converted debug/trace/warn/info checks to lambda based logging APIs.

Resolves #8646

Signed-off-by: Abdul Muneer Kolarkunnu <muneer.kolarkunnu@netapp.com>

akolarkunnu commented on server/src/main/java/org/opensearch/transport/TransportService.java:
```diff
@@ -1318,14 +1323,11 @@ private void checkForTimeout(long requestId) {
             sourceNode = null;
         }
         // call tracer out of lock
-        if (tracerLog.isTraceEnabled() == false) {
```
Added it back

sandeshkr419 commented:
@akolarkunnu Did you get a chance to get to profiling and benchmarking?
It will be real nice to get these changes ahead of 3.0 release actually.
